tools/cldr-apps/src/main/java/org/unicode/cldr/util/CLDRCacheDir.java

```diff
@@ -0,0 +1,191 @@
+        Instant latestGood = now.minus(CLDRConfig.getInstance().getProperty(propName, expireDays), ChronoUnit.DAYS);
+        return latestGood;
+    }
+
+    private final CacheType type;
+
+    private File file;
+
+    CLDRCacheDir(CacheType type) {
+        this.type = type;
+        CACHE_HELPER.assertOK(); // throw any exception needed
+        this.file = getFile(this.type);
+    }
+
+    /**
+     * Construct a new Cache Dir handle using the specified type
+     * @param type
+     * @return
+     */
+    public static CLDRCacheDir getInstance(CacheType type) {
+        return new CLDRCacheDir(type);
+    }
+
+    private static File getFile(CacheType type) {
+        return new File(CACHE_HELPER.rootDir, type.name());
+    }
+
+    /**
+     * Get the File pertaining to this cache entry.
+     * @return
+     */
+    public File getDir() {
+        return file;
+    }
+
+    /**
+     * Return the file, as an empty dir.
+     * This is the function most callers will want to use.
+     * @return
+     */
+    public File getEmptyDir() {
+        getDir().mkdirs();
+        return getDir();
+    }
+
+    public CacheType getType() {
+        return type;
+    }
+
+    /**
+     * Delete the specified cache entry.
+     * @throws IOException
+     */
+    void remove() throws IOException {
+        FileUtils.deleteDirectory(getDir());
+    }
+
+    /**
+     * Remove all cache entries.
+     * Leaves the cache directory itself.
+     * @throws IOException
+     */
+    static void removeAll() throws IOException {
+        for (final CacheType t : CacheType.values()) {
+            FileUtils.deleteDirectory(getFile(t));
+        }
+    }
+
+    static final class CacheHelper {
+        File rootDir = null;
+        Throwable err = null;
+
+        CacheHelper() {
+            try {
+                final CLDRConfig config = CLDRConfig.getInstance();
+                if(config instanceof CLDRConfigImpl) {
+                    File homeFile = CLDRConfigImpl.getInstance().homeFile;
+                    if(homeFile == null) {
+                        err = new IllegalArgumentException("CLDRConfigImpl present but homeFile not set.");
+                    } else {
+                        rootDir = new File(homeFile, CACHE_SUBDIR);
+                    }
+                } else {
+                    rootDir = Files.createTempDir();
+                }
+                System.out.println("CLDRCacheDir: " + rootDir.getAbsolutePath());
+                setup();
+            } catch(Throwable t) {
+                err = t;
+            }
+        }
+
+        void assertOK() throws IllegalArgumentException {
+            if(err != null) {
+                throw new IllegalArgumentException("Not setup OK", err);
+            }
+        }
+
+        /**
+         * Make sure the cache directory is setup.
+         * @throws IOException
+         */
+        void setup() throws IOException {
+            if(rootDir != null) {
```

Comment:
what if rootDir is null? warning in that case?

Reply:
@btangmu I'm going to take some time to improve the apidocs around SurveyTool startup.

Reply:
done